The job **`x86_64-gnu-llvm-19`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/15211082691/job/42785209389) [(plain)](https://github.com/rust-lang/rust/commit/4e9a021eea782296d1fa007fb1a857a36e198e6c/checks/42785209389/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#18 exporting to docker image format
#18 sending tarball 21.8s done
#18 DONE 30.0s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-19]
[CI_JOB_NAME=x86_64-gnu-llvm-19]
debug: `DISABLE_CI_RUSTC_IF_INCOMPATIBLE` configured.
---
sccache: Listening on address 127.0.0.1:4226
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-19', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'build.print-step-timings', '--enable-verbose-tests', '--set', 'build.metrics', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--set', 'gcc.download-ci-gcc=true', '--enable-new-symbol-mangling']
configure: build.build          := x86_64-unknown-linux-gnu
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-19/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
---
diff of stderr:

21    |
22    = help: the trait `FromIterator<()>` is not implemented for `String`
23    = help: the following other types implement trait `FromIterator<A>`:
+              `String` implements `FromIterator<&Char>`
24              `String` implements `FromIterator<&char>`
25              `String` implements `FromIterator<&str>`
26              `String` implements `FromIterator<Box<str, A>>`

+              `String` implements `FromIterator<Char>`
27              `String` implements `FromIterator<Cow<'_, str>>`
28              `String` implements `FromIterator<String>`
29              `String` implements `FromIterator<char>`
---
To only update this specific test, also pass `--test-args suggestions/semi-suggestion-when-stmt-and-expr-span-equal.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/suggestions/semi-suggestion-when-stmt-and-expr-span-equal.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--check-cfg" "cfg(test,FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/suggestions/semi-suggestion-when-stmt-and-expr-span-equal" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers"
stdout: none
--- stderr -------------------------------
error[E0308]: mismatched types
##[error]  --> /checkout/tests/ui/suggestions/semi-suggestion-when-stmt-and-expr-span-equal.rs:13:5
   |
LL |   fn foo() -> String {
   |               ------ expected `String` because of return type
LL |       let mut list = {
   |  ____________________-
LL | |         println!();
LL | |     };
   | |_____- this block is missing a tail expression
LL |       list //~ ERROR mismatched types
   |       ^^^^ expected `String`, found `()`

error[E0277]: a value of type `String` cannot be built from an iterator over elements of type `()`
##[error]  --> /checkout/tests/ui/suggestions/semi-suggestion-when-stmt-and-expr-span-equal.rs:23:20
   |
LL |         .collect::<String>(); //~ ERROR E0277
   |          -------   ^^^^^^ value of type `String` cannot be built from `std::iter::Iterator<Item=()>`
   |          |
   |          required by a bound introduced by this call
   |
   = help: the trait `FromIterator<()>` is not implemented for `String`
   = help: the following other types implement trait `FromIterator<A>`:
             `String` implements `FromIterator<&Char>`
             `String` implements `FromIterator<&char>`
             `String` implements `FromIterator<&str>`
             `String` implements `FromIterator<Box<str, A>>`
             `String` implements `FromIterator<Char>`
             `String` implements `FromIterator<Cow<'_, str>>`
             `String` implements `FromIterator<String>`
             `String` implements `FromIterator<char>`
note: the method call chain might not have had the expected associated types
  --> /checkout/tests/ui/suggestions/semi-suggestion-when-stmt-and-expr-span-equal.rs:20:10
   |
LL |       String::new()
   |       ------------- this expression has type `String`
LL |           .chars()
   |            ------- `Iterator::Item` is `char` here
LL |           .filter(|x| !x.is_whitespace())
   |            ------------------------------ `Iterator::Item` remains `char` here
LL |           .map(|x| {
   |  __________^
LL | |             println!("Child spawned with the size: {}", x);
LL | |         })
   | |__________^ `Iterator::Item` changed to `()` here
note: required by a bound in `collect`
  --> /rustc/FAKE_PREFIX/library/core/src/iter/traits/iterator.rs:1988:5

error: aborting due to 2 previous errors

```

</details>

--- a/library/alloc/src/string.rs
+++ b/library/alloc/src/string.rs
@@ -2374,6 +2374,28 @@ impl<'a> FromIterator<Cow<'a, str>> for String {
 }
 
 #[cfg(not(no_global_oom_handling))]
+#[unstable(feature = "ascii_char", issue = "110998")]
+impl FromIterator<core::ascii::Char> for String {
+    fn from_iter<T: IntoIterator<Item = core::ascii::Char>>(iter: T) -> Self {
+        let buf = iter.into_iter().map(core::ascii::Char::to_u8).collect();
+        // SAFETY: `buf` is guaranteed to be valid UTF-8 because the `core::ascii::Char` type
+        // only contains ASCII values (0x00-0x7F), which are valid UTF-8.
+        unsafe { String::from_utf8_unchecked(buf) }
+    }
+}
+
+#[cfg(not(no_global_oom_handling))]
+#[unstable(feature = "ascii_char", issue = "110998")]
+impl<'a> FromIterator<&'a core::ascii::Char> for String {
+    fn from_iter<T: IntoIterator<Item = &'a core::ascii::Char>>(iter: T) -> Self {
+        let buf = iter.into_iter().copied().map(core::ascii::Char::to_u8).collect();
+        // SAFETY: `buf` is guaranteed to be valid UTF-8 because the `core::ascii::Char` type
+        // only contains ASCII values (0x00-0x7F), which are valid UTF-8.
+        unsafe { String::from_utf8_unchecked(buf) }
+    }
+}
+
+#[cfg(not(no_global_oom_handling))]
 #[stable(feature = "rust1", since = "1.0.0")]
 impl Extend<char> for String {
     fn extend<I: IntoIterator<Item = char>>(&mut self, iter: I) {
@@ -3196,6 +3218,14 @@ impl<'a, 'b> FromIterator<&'b str> for Cow<'a, str> {
 #[stable(feature = "cow_str_from_iter", since = "1.12.0")]
 impl<'a> FromIterator<String> for Cow<'a, str> {
     fn from_iter<I: IntoIterator<Item = String>>(it: I) -> Cow<'a, str> {
+        Cow::Owned(FromIterator::from_iter(it))
+    }
+}
+
+#[cfg(not(no_global_oom_handling))]
+#[unstable(feature = "ascii_char", issue = "110998")]
+impl<'a> FromIterator<core::ascii::Char> for Cow<'a, str> {
+    fn from_iter<T: IntoIterator<Item = core::ascii::Char>>(it: T) -> Self {
         Cow::Owned(FromIterator::from_iter(it))
     }
 }

--- a/tests/ui/suggestions/semi-suggestion-when-stmt-and-expr-span-equal.stderr
+++ b/tests/ui/suggestions/semi-suggestion-when-stmt-and-expr-span-equal.stderr
@@ -21,9 +21,11 @@ LL |         .collect::<String>();
    |
    = help: the trait `FromIterator<()>` is not implemented for `String`
    = help: the following other types implement trait `FromIterator<A>`:
+             `String` implements `FromIterator<&Char>`
              `String` implements `FromIterator<&char>`
              `String` implements `FromIterator<&str>`
              `String` implements `FromIterator<Box<str, A>>`
+             `String` implements `FromIterator<Char>`
              `String` implements `FromIterator<Cow<'_, str>>`
              `String` implements `FromIterator<String>`
              `String` implements `FromIterator<char>`

